clippy_lints/src/redundant_test_prefix.rs

```diff
@@ -0,0 +1,168 @@
+        _decl: &FnDecl<'_>,
+        body: &'tcx Body<'_>,
+        _span: Span,
+        fn_def_id: LocalDefId,
+    ) {
+        // Ignore methods and closures.
+        let FnKind::ItemFn(ref ident, ..) = kind else {
+            return;
+        };
+
+        // Skip the lint if the function is within a macro expansion.
+        if ident.span.from_expansion() {
+            return;
+        }
+
+        // Skip if the function name does not start with `test_`.
+        if !ident.as_str().starts_with("test_") {
+            return;
+        }
+
+        if is_test_function(cx.tcx, cx.tcx.local_def_id_to_hir_id(fn_def_id)) {
+            let msg = "redundant `test_` prefix in test function name";
+            let mut non_prefixed = ident.as_str().trim_start_matches("test_").to_string();
+
+            if is_invalid_ident(&non_prefixed) {
+                // If the prefix-trimmed name is not a valid function name, do not provide an
+                // automatic fix, just suggest renaming the function.
+                span_lint_and_help(
+                    cx,
+                    REDUNDANT_TEST_PREFIX,
+                    ident.span,
+                    msg,
+                    None,
+                    "consider function renaming (just removing `test_` prefix will produce invalid function name)",
+                );
+            } else if name_conflicts(cx, body, &non_prefixed) {
+                // If `non_prefixed` conflicts with another function in the same module/scope,
+                // do not provide an automatic fix, but still emit a fix suggestion.
+                non_prefixed.push_str("_works");
+                span_lint_and_sugg(
+                    cx,
+                    REDUNDANT_TEST_PREFIX,
+                    ident.span,
+                    msg,
+                    "consider function renaming (just removing `test_` prefix will cause a name conflict)",
+                    non_prefixed,
+                    Applicability::MaybeIncorrect,
+                );
+            } else {
+                // If `non_prefixed` is a valid identifier and does not conflict with another function,
+                // so we can suggest an auto-fix.
+                span_lint_and_sugg(
+                    cx,
+                    REDUNDANT_TEST_PREFIX,
+                    ident.span,
+                    msg,
+                    "consider removing the `test_` prefix",
+                    non_prefixed,
+                    Applicability::MaybeIncorrect,
+                );
+            }
+        }
+    }
+}
+
+/// Checks whether removal of the `_test` prefix from the function name will cause a name conflict.
+///
+/// There should be no other function with the same name in the same module/scope. Also, there
+/// should not be any function call with the same name within the body of the function, to avoid
+/// recursion.
+fn name_conflicts<'tcx>(cx: &LateContext<'tcx>, body: &'tcx Body<'_>, fn_name: &str) -> bool {
```

Comment:
Given the number of comparisons you are going to make, you should maybe `Symbol::intern()` this name. Moreover, you seem to be interning it later to check for validity, so this might be shared.